src/strategy.rs

```diff
@@ -261,15 +271,400 @@ pub fn hostname<S: BuildHasher>(parameters: Option<HashMap<String, String, S>>)
+                                    return false;
+                                }
+                            }
+                            true
+                        })
+                        .unwrap_or(false)
+                })
+            }
+        }
+    }
+}
+
+fn _compile_constraints(constraints: Vec<Constraint>) -> Vec<Evaluate> {
+    constraints
+        .into_iter()
+        .map(|constraint| {
+            let (context_name, expression) = (constraint.context_name, constraint.expression);
+            match context_name.as_str() {
+                "appName" => {
+                    _compile_constraint_string(expression, |context| Some(&context.app_name))
+                }
+                "environment" => {
+                    _compile_constraint_string(expression, |context| Some(&context.environment))
+                }
+                "remoteAddress" => {
+                    _compile_constraint_host(expression, |context| context.remote_address.as_ref())
+                }
+                "sessionId" => {
+                    _compile_constraint_string(expression, |context| context.session_id.as_ref())
+                }
+                "userId" => {
+                    _compile_constraint_string(expression, |context| context.user_id.as_ref())
+                }
+                _ => _compile_constraint_string(expression, move |context| {
+                    context.properties.get(&context_name)
+                }),
+            }
+        })
+        .collect()
+}
+
+/// This function is a strategy decorator which compiles to nothing when
+/// there are no constraints, or to a constraint evaluating test if there are.
+pub fn constrain<S: Fn(Option<HashMap<String, String>>) -> Evaluate + Sync + Send + 'static>(
+    constraints: Option<Vec<Constraint>>,
+    strategy: &S,
+    parameters: Option<HashMap<String, String>>,
+) -> Evaluate {
+    let compiled_strategy = strategy(parameters);
+    match constraints {
+        None => {
+            trace!("constrain: no constraints, bypassing");
+            compiled_strategy
+        }
+        Some(constraints) => {
+            if constraints.is_empty() {
+                trace!("constrain: empty constraints list, bypassing");
+                compiled_strategy
+            } else {
+                trace!("constrain: compiling constraints list {:?}", constraints);
+                let constraints = _compile_constraints(constraints);
+                // Create a closure that will evaluate against the context.
+                Box::new(move |context| {
+                    // Check every constraint; if all match, permit
+                    for constraint in &constraints {
+                        if !constraint(&context) {
+                            return false;
+                        }
+                    }
+                    compiled_strategy(context)
+                })
+            }
+        }
+    }
+}
+
 #[cfg(test)]
 mod tests {
```

Comment:
Should the `tests` module be moved out into its own fil?

Reply:
pretty sure not; this is very much close to the bone unit tests, having them adjacent to the code is the norm. I know the tokio ecosystem doesn't do that :).

Reply:
As mentioned in my other comment, this should be effectively the same, just not keeping the tests module inline. (So to be clear, I'm talking about moving it to `src/strategy/tests.rs`, not <code>tests/<i>something</i>.rs</code>.)